File: backend/routers/game/locations.py
"""
Location management routes.

Endpoints for listing locations, traveling, updating labels, and getting location messages.
"""

import logging
from typing import Optional

import crud
import schemas
from database import get_db
from dependencies import (
    RequestIdentity,
    get_request_identity,
)
from domain.services.access_control import AccessControl
from fastapi import APIRouter, Depends, HTTPException
from services.location_service import LocationService
from sqlalchemy.ext.asyncio import AsyncSession

logger = logging.getLogger("GameRouter.Locations")

router = APIRouter()


@router.get("/{world_id}/locations", response_model=list[schemas.Location])
async def list_locations(
    world_id: int,
    discovered_only: bool = True,
    db: AsyncSession = Depends(get_db),
    identity: RequestIdentity = Depends(get_request_identity),
):
    """
    List all locations in the world.

    By default, only returns discovered locations.
    """
    world = await crud.get_world(db, world_id)
    if not world:
        raise HTTPException(status_code=404, detail="World not found")
    AccessControl.raise_if_no_access(identity.user_id, identity.role, world.owner_id)

    locations = await crud.get_locations(db, world_id)

    if discovered_only:
        locations = [loc for loc in locations if loc.is_discovered]

    return locations


@router.get("/{world_id}/locations/current", response_model=schemas.Location)
async def get_current_location(
    world_id: int,
    db: AsyncSession = Depends(get_db),
    identity: RequestIdentity = Depends(get_request_identity),
):
    """Get the player's current location."""
    world = await crud.get_world(db, world_id)
    if not world:
        raise HTTPException(status_code=404, detail="World not found")
    AccessControl.raise_if_no_access(identity.user_id, identity.role, world.owner_id)

    player_state = await crud.get_player_state(db, world_id)
    if not player_state or not player_state.current_location_id:
        raise HTTPException(status_code=404, detail="No current location")

    location = await crud.get_location(db, player_state.current_location_id)
    if not location:
        raise HTTPException(status_code=404, detail="Location not found")

    return location


@router.post("/{world_id}/locations/{location_id}/travel")
async def travel_to_location(
    world_id: int,
    location_id: int,
    db: AsyncSession = Depends(get_db),
    identity: RequestIdentity = Depends(get_request_identity),
):
    """
    Travel to a different location.

    This is equivalent to submitting a travel action.
    """
    world = await crud.get_world(db, world_id)
    if not world:
        raise HTTPException(status_code=404, detail="World not found")
    AccessControl.raise_if_no_access(identity.user_id, identity.role, world.owner_id)

    # Verify location exists and belongs to this world
    location = await crud.get_location(db, location_id)
    if not location or location.world_id != world_id:
        raise HTTPException(status_code=404, detail="Location not found")

    # Update current location in database
    await crud.set_current_location(db, world_id, location_id)

    # Update current room in _state.json for filesystem sync
    room_key = LocationService.location_to_room_key(location.name)
    LocationService.set_current_room(world.name, room_key)

    logger.info(f"Traveled to location {location.display_name or location.name}")

    return {
        "status": "traveled",
        "destination": location.display_name or location.name,
        "location_id": location_id,
    }


@router.patch("/{world_id}/locations/{location_id}", response_model=schemas.Location)
async def update_location_label(
    world_id: int,
    location_id: int,
    update: schemas.LocationUpdate,
    db: AsyncSession = Depends(get_db),
    identity: RequestIdentity = Depends(get_request_identity),
):
    """
    Update a location's user-assigned label.

    This is the only modification users can make to locations.
    """
    world = await crud.get_world(db, world_id)
    if not world:
        raise HTTPException(status_code=404, detail="World not found")
    AccessControl.raise_if_no_access(identity.user_id, identity.role, world.owner_id)

    location = await crud.update_location_label(db, location_id, update.label)
    if not location:
        raise HTTPException(status_code=404, detail="Location not found")

    return location


@router.get("/{world_id}/locations/{location_id}/messages")
async def get_location_messages(
    world_id: int,
    location_id: int,
    limit: int = 50,
    since_id: Optional[int] = None,
    db: AsyncSession = Depends(get_db),
    identity: RequestIdentity = Depends(get_request_identity),
):
    """
    Get messages (chat history) for a specific location.

    Used to view past interactions at that location.
    """
    world = await crud.get_world(db, world_id)
    if not world:
        raise HTTPException(status_code=404, detail="World not found")
    AccessControl.raise_if_no_access(identity.user_id, identity.role, world.owner_id)

    location = await crud.get_location(db, location_id)
    if not location or not location.room_id:
        raise HTTPException(status_code=404, detail="Location not found")

    # Use existing message query logic
    if since_id:
        messages = await crud.get_messages_since(db, location.room_id, since_id=since_id, limit=limit)
    else:
        messages = await crud.get_messages(db, location.room_id)
        # Apply limit manually since get_messages doesn't support it
        messages = messages[-limit:] if len(messages) > limit else messages

    # Format messages consistently with poll endpoint
    formatted_messages = [
        {
            "id": m.id,
            "content": m.content,
            "role": m.role,
            "agent_id": m.agent_id,
            "agent_name": m.agent.name if m.agent else None,
            "thinking": m.thinking,
            "timestamp": m.timestamp.isoformat() if m.timestamp else None,
            "image_data": m.image_data,
            "image_media_type": m.image_media_type,
        }
        for m in messages
    ]

    return {"messages": formatted_messages}

File: backend-ts/src/http/routes/game/locations.ts
/**
 * Location routes — port of `backend/routers/game/locations.py`.
 *
 * The map, as the left-hand panel sees it: which places are known, where the
 * player is, moving between them, renaming one, and reading back a place's
 * transcript.
 *
 * Registration order matters here for one pair. `GET .../locations/current` and
 * `GET .../locations/{location_id}/messages` do not collide, but
 * `.../locations/current` *would* collide with a `GET .../locations/{id}` route
 * if one were ever added — there is none in Python, and adding one would need
 * `current` to stay first.
 */

import { Hono } from 'hono'

import { getLocation, getLocations, updateLocationLabel } from '../../../crud/locations'
import { getMessages, getMessagesSince } from '../../../crud/messages'
import { getPlayerState, setCurrentLocation } from '../../../crud/player-state'
import { getLogger } from '../../../infrastructure/logging/logger'
import { LocationUpdate, toLocation } from '../../../schemas/game'
import { RoomMappingService } from '../../../services/room-mapping'
import { HttpError } from '../../errors'
import type { AppState } from '../../state'
import type { AppEnv } from '../../types'
import {
  boolQueryParam,
  intPathParam,
  intQueryParam,
  intQueryParamOr,
  parseBody,
  requireWorld,
  toPollMessage,
} from './shared'

const logger = getLogger('GameRouter.Locations')

export function createLocationRoutes(state: AppState): Hono<AppEnv> {
  const routes = new Hono<AppEnv>()

  /**
   * The world's locations, discovered ones only by default.
   *
   * The filter is applied in memory rather than in SQL, as in Python — the map
   * of a world is tens of rows, and `?discovered_only=false` (used by the world
   * inspector) wants the same query.
   */
  routes.get('/worlds/:world_id/locations', (c) => {
    const worldId = intPathParam(c, 'world_id')
    const discoveredOnly = boolQueryParam(c, 'discovered_only', true)
    requireWorld(state, c, worldId)

    const all = getLocations(state.db, worldId)
    const visible = discoveredOnly ? all.filter((location) => location.isDiscovered) : all
    return c.json(visible.map(toLocation))
  })

  /** Where the player is standing. 404 while they are nowhere (onboarding). */
  routes.get('/worlds/:world_id/locations/current', (c) => {
    const worldId = intPathParam(c, 'world_id')
    requireWorld(state, c, worldId)

    const playerState = getPlayerState(state.db, worldId)
    if (!playerState?.currentLocationId) throw new HttpError(404, 'No current location')

    const location = getLocation(state.db, playerState.currentLocationId)
    if (!location) throw new HttpError(404, 'Location not found')

    return c.json(toLocation(location))
  })

  /**
   * Move the player to another location.
   *
   * Two writes, in this order and both required: the database row (which also
   * flips `is_current` and marks the destination discovered, inside
   * `setCurrentLocation`), and `_state.json`'s `current_room`, which is what the
   * agent-side tools read to decide whose room they are acting in. Leaving the
   * second out is invisible until the next turn narrates the wrong place.
   *
   * Note this endpoint does *not* run a turn. It is the map-click path; the
   * narrated version of travelling is the Action Manager's `travel` tool.
   */
  routes.post('/worlds/:world_id/locations/:location_id/travel', (c) => {
    const worldId = intPathParam(c, 'world_id')
    const locationId = intPathParam(c, 'location_id')
    const world = requireWorld(state, c, worldId)

    const location = getLocation(state.db, locationId)
    // The world check is what stops a location id from one world being used to
    // move the player inside another.
    if (!location || location.worldId !== worldId) {
      throw new HttpError(404, 'Location not found')
    }

    setCurrentLocation(state.db, worldId, locationId)
    state.services.rooms.setCurrentRoom(
      world.name,
      RoomMappingService.locationToRoomKey(location.name),
    )

    logger.info(`Traveled to location ${location.displayName || location.name}`)

    return c.json({
      status: 'traveled',
      destination: location.displayName || location.name,
      location_id: locationId,
    })
  })

  /**
   * Set a location's user label — the only field a player may edit on the map.
   *
   * **The location is not checked against the world.** Python looks the world up
   * for the access check and then updates by location id alone, so an
   * authenticated caller can relabel any location in any world. Reproduced
   * rather than tightened: labels are cosmetic, and the parity harness compares
   * statuses.
   */
  routes.patch('/worlds/:world_id/locations/:location_id', async (c) => {
    const worldId = intPathParam(c, 'world_id')
    const locationId = intPathParam(c, 'location_id')
    const update = await parseBody(c, LocationUpdate)
    requireWorld(state, c, worldId)

    const location = updateLocationLabel(state.db, locationId, update.label)
    if (!location) throw new HttpError(404, 'Location not found')

    return c.json(toLocation(location))
  })

  /**
   * A location's transcript.
   *
   * `since_id` switches to the incremental query; without it the *last* `limit`
   * messages are returned, trimmed in memory because `getMessages` is
   * deliberately unbounded. The message shape is the poll endpoint's, not
   * `schemas.Message` — see {@link toPollMessage}.
   */
  routes.get('/worlds/:world_id/locations/:location_id/messages', (c) => {
    const worldId = intPathParam(c, 'world_id')
    const locationId = intPathParam(c, 'location_id')
    const limit = intQueryParamOr(c, 'limit', 50)
    const sinceId = intQueryParam(c, 'since_id')
    requireWorld(state, c, worldId)

    const location = getLocation(state.db, locationId)
    if (!location?.roomId) throw new HttpError(404, 'Location not found')

    const messages = sinceId
      ? getMessagesSince(state.db, location.roomId, sinceId, limit)
      : trimToLast(getMessages(state.db, location.roomId), limit)

    return c.json({ messages: messages.map(toPollMessage) })
  })

  return routes
}

function trimToLast<T>(items: T[], limit: number): T[] {
  return items.length > limit ? items.slice(-limit) : items
}
